Contexto
Se você gosta de comer cereal, faça um favor a si mesmo e evite este conjunto de dados a todo custo. Depois de ver esses dados, nunca mais será a mesma coisa para mim comer Fruity Pebbles.

#### Contente

Campos no conjunto de dados:

- Name: Name of cereal

- mfr: Manufacturer of cereal

- A = American Home Food Products;

- G = General Mills

- K = Kelloggs

- N = Nabisco

- P = Post

- Q = Quaker Oats

- R = Ralston Purina

#### Tipo:

  - cold

  - hot

- calories: calories per serving

- protein: grams of protein

- fat: grams of fat

- sodium: milligrams of sodium

- fiber: grams of dietary fiber

- carbo: grams of complex carbohydrates

- sugars: grams of sugars

- potass: milligrams of potassium

- vitamins: vitamins and minerals - 0, 25, or 100, indicating the typical percentage of FDA recommended

- shelf: display shelf (1, 2, or 3, counting from the floor)

- weight: weight in ounces of one serving

- cups: number of cups in one serving

- rating: a rating of the cereals (Possibly from Consumer Reports?)


### Importando base

In [46]:
import pandas as pd

- Exercício 1 — head() e tail()

Carregue o dataset cereal.csv e imprima:

As primeiras 5 linhas

As últimas 3 linhas

In [20]:
#Carregando o dataset
df = pd.read_csv('cereal.csv')


In [21]:
df.head()

,name,mfr,type,calories,protein,fat,sodium,fiber,carbo,sugars,potass,vitamins,shelf,weight,cups,rating
0,100% Bran,N,C,70,4,1,130,10.0,5.0,6,280,25,3,1.0,0.33,68.402973
1,100% Natural Bran,Q,C,120,3,5,15,2.0,8.0,8,135,0,3,1.0,1.00,33.983679
2,All-Bran,K,C,70,4,1,260,9.0,7.0,5,320,25,3,1.0,0.33,59.425505
3,All-Bran with Extra Fiber,K,C,50,4,0,140,14.0,8.0,0,330,25,3,1.0,0.50,93.704912
4,Almond Delight,R,C,110,2,2,200,1.0,14.0,8,-1,25,3,1.0,0.75,34.384843


In [22]:
df.tail(3)

,name,mfr,type,calories,protein,fat,sodium,fiber,carbo,sugars,potass,vitamins,shelf,weight,cups,rating
74,Wheat Chex,R,C,100,3,1,230,3.0,17.0,3,115,25,1,1.0,0.67,49.787445
75,Wheaties,G,C,100,3,1,200,3.0,17.0,3,110,25,1,1.0,1.00,51.592193
76,Wheaties Honey Gold,G,C,110,2,1,200,1.0,16.0,8,60,25,1,1.0,0.75,36.187559


- Exercício 2 — shape 🎯

Com o mesmo dataset, imprima:

Quantas linhas tem o dataset

Quantas colunas tem o dataset

Uma mensagem: "O dataset tem X cereais e Y características"

In [28]:
#Quantas linhas há no dataframe
print(f"Linhas: {df.shape[0]}")

Linhas: 77


In [27]:
print(f"Colunas: {df.shape[1]}")

Colunas: 16


In [30]:
linhas, colunas = df.shape
print(f"Quantidade de linhas: {linhas}\nQuantidade de colunas: {colunas}")

Quantidade de linhas: 77
Quantidade de colunas: 16


- Exercício 3 — columns 

Imprima todos os nomes das colunas e use for para imprimir cada coluna com seu número:

1 - name

2 - calories

3 - protein
...

In [35]:
columns = df.columns

In [38]:
print(columns)

Index(['name', 'mfr', 'type', 'calories', 'protein', 'fat', 'sodium', 'fiber',
       'carbo', 'sugars', 'potass', 'vitamins', 'shelf', 'weight', 'cups',
       'rating'],
      dtype='object')


In [42]:
for i in range(len(columns)): 
    print(f"{i + 1} - {columns[i]}")

1 - name
2 - mfr
3 - type
4 - calories
5 - protein
6 - fat
7 - sodium
8 - fiber
9 - carbo
10 - sugars
11 - potass
12 - vitamins
13 - shelf
14 - weight
15 - cups
16 - rating


- Exercício 4 — info() 🎯

Use df.info() e responda nos comentários do código:

Quantas colunas numéricas tem?

  R: 13 Colunas numeros (5 float / 8 int)

Quantas colunas de texto tem?

  R: Há 3 Colunas. 

Tem algum valor faltando?
  R: Não

In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77 entries, 0 to 76
Data columns (total 16 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   name      77 non-null     object 
 1   mfr       77 non-null     object 
 2   type      77 non-null     object 
 3   calories  77 non-null     int64  
 4   protein   77 non-null     int64  
 5   fat       77 non-null     int64  
 6   sodium    77 non-null     int64  
 7   fiber     77 non-null     float64
 8   carbo     77 non-null     float64
 9   sugars    77 non-null     int64  
 10  potass    77 non-null     int64  
 11  vitamins  77 non-null     int64  
 12  shelf     77 non-null     int64  
 13  weight    77 non-null     float64
 14  cups      77 non-null     float64
 15  rating    77 non-null     float64
dtypes: float64(5), int64(8), object(3)
memory usage: 9.8+ KB


- Use df.describe() com formatação de 2 casas decimais e responda nos comentários do código:

Qual cereal tem mais calorias?

Qual é a média de proteínas?

Qual é a mediana de açúcar?

In [51]:
with pd.option_context("float_format", "{:.2f}".format):
    display(df.describe())

,calories,protein,fat,sodium,fiber,carbo,sugars,potass,vitamins,shelf,weight,cups,rating
count,77.00,77.00,77.00,77.00,77.00,77.00,77.00,77.00,77.00,77.00,77.00,77.00,77.00
mean,106.88,2.55,1.01,159.68,2.15,14.60,6.92,96.08,28.25,2.21,1.03,0.82,42.67
std,19.48,1.09,1.01,83.83,2.38,4.28,4.44,71.29,22.34,0.83,0.15,0.23,14.05
min,50.00,1.00,0.00,0.00,0.00,-1.00,-1.00,-1.00,0.00,1.00,0.50,0.25,18.04
25%,100.00,2.00,0.00,130.00,1.00,12.00,3.00,40.00,25.00,1.00,1.00,0.67,33.17
50%,110.00,3.00,1.00,180.00,2.00,14.00,7.00,90.00,25.00,2.00,1.00,0.75,40.40
75%,110.00,3.00,2.00,210.00,3.00,17.00,11.00,120.00,25.00,3.00,1.00,1.00,50.83
max,160.00,6.00,5.00,320.00,14.00,23.00,15.00,330.00,100.00,3.00,1.50,1.50,93.70


In [57]:
df["calories"].max()

160

In [ ]:
#Sem muitas casas decimais
print(f"{df["protein"].mean():.2f}")

2.55


In [64]:
#Com cada decimal
df["protein"].mean()


2.5454545454545454

In [65]:
df["sugars"].median()

7.0

- Exercício 6 — describe(exclude="number") 🎯

Use df.describe(exclude="number") e responda nos comentários do código:

Quantos fabricantes diferentes existem?
        - 7 

Qual é o fabricante mais comum?

        - Kallogs

Quantos tipos de cereal existem?
        - 77


In [66]:
df.describe(exclude="number")

,name,mfr,type
count,77,77,77
unique,77,7,2
top,100% Bran,K,C
freq,1,23,74


In [72]:
df["mfr"].unique()

array(['N', 'Q', 'K', 'R', 'G', 'P', 'A'], dtype=object)

0                     100% Bran
1             100% Natural Bran
2                      All-Bran
3     All-Bran with Extra Fiber
4                Almond Delight
                ...            
72                      Triples
73                         Trix
74                   Wheat Chex
75                     Wheaties
76          Wheaties Honey Gold
Name: name, Length: 77, dtype: object

In [78]:
df

,name,mfr,type,calories,protein,fat,sodium,fiber,carbo,sugars,potass,vitamins,shelf,weight,cups,rating
0,100% Bran,N,C,70,4,1,130,10.0,5.0,6,280,25,3,1.0,0.33,68.402973
1,100% Natural Bran,Q,C,120,3,5,15,2.0,8.0,8,135,0,3,1.0,1.00,33.983679
2,All-Bran,K,C,70,4,1,260,9.0,7.0,5,320,25,3,1.0,0.33,59.425505
3,All-Bran with Extra Fiber,K,C,50,4,0,140,14.0,8.0,0,330,25,3,1.0,0.50,93.704912
4,Almond Delight,R,C,110,2,2,200,1.0,14.0,8,-1,25,3,1.0,0.75,34.384843
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
72,Triples,G,C,110,2,1,250,0.0,21.0,3,60,25,3,1.0,0.75,39.106174
73,Trix,G,C,110,1,1,140,0.0,13.0,12,25,25,2,1.0,1.00,27.753301
74,Wheat Chex,R,C,100,3,1,230,3.0,17.0,3,115,25,1,1.0,0.67,49.787445
75,Wheaties,G,C,100,3,1,200,3.0,17.0,3,110,25,1,1.0,1.00,51.592193


In [ ]:

df["type"].unique()

array(['C', 'H'], dtype=object)

- Exercício 7 — isnull() 🎯

Use df.isnull().sum() e responda nos comentários do código:

Tem alguma coluna com valores faltando?

Qual coluna tem mais valores faltando?

In [80]:
df.isnull().sum()

name        0
mfr         0
type        0
calories    0
protein     0
fat         0
sodium      0
fiber       0
carbo       0
sugars      0
potass      0
vitamins    0
shelf       0
weight      0
cups        0
rating      0
dtype: int64

- Exercício 8 — Combinando tudo 🎯

Use os métodos que aprendeu para responder:

Quantas linhas e colunas tem o dataset? (shape)

Quais são as primeiras 3 linhas? (head)

Tem valores faltando? (isnull)

Qual é a média de calorias? (describe)

In [83]:
#Quantidade de linhas e colunas:
df.shape

(77, 16)

In [84]:
df.head(3)

,name,mfr,type,calories,protein,fat,sodium,fiber,carbo,sugars,potass,vitamins,shelf,weight,cups,rating
0,100% Bran,N,C,70,4,1,130,10.0,5.0,6,280,25,3,1.0,0.33,68.402973
1,100% Natural Bran,Q,C,120,3,5,15,2.0,8.0,8,135,0,3,1.0,1.00,33.983679
2,All-Bran,K,C,70,4,1,260,9.0,7.0,5,320,25,3,1.0,0.33,59.425505


In [87]:
df.isnull().sum()

name        0
mfr         0
type        0
calories    0
protein     0
fat         0
sodium      0
fiber       0
carbo       0
sugars      0
potass      0
vitamins    0
shelf       0
weight      0
cups        0
rating      0
dtype: int64

In [101]:
print(f"{df["calories"].mean():.2f}")

106.88


- Exercício 9 — head() com número específico

Imprima:

As primeiras 10 linhas

As últimas 7 linhas

Compare o primeiro e o último cereal do dataset

In [102]:
df.head(10)

,name,mfr,type,calories,protein,fat,sodium,fiber,carbo,sugars,potass,vitamins,shelf,weight,cups,rating
0,100% Bran,N,C,70,4,1,130,10.0,5.0,6,280,25,3,1.00,0.33,68.402973
1,100% Natural Bran,Q,C,120,3,5,15,2.0,8.0,8,135,0,3,1.00,1.00,33.983679
2,All-Bran,K,C,70,4,1,260,9.0,7.0,5,320,25,3,1.00,0.33,59.425505
3,All-Bran with Extra Fiber,K,C,50,4,0,140,14.0,8.0,0,330,25,3,1.00,0.50,93.704912
4,Almond Delight,R,C,110,2,2,200,1.0,14.0,8,-1,25,3,1.00,0.75,34.384843
5,Apple Cinnamon Cheerios,G,C,110,2,2,180,1.5,10.5,10,70,25,1,1.00,0.75,29.509541
6,Apple Jacks,K,C,110,2,0,125,1.0,11.0,14,30,25,2,1.00,1.00,33.174094
7,Basic 4,G,C,130,3,2,210,2.0,18.0,8,100,25,3,1.33,0.75,37.038562
8,Bran Chex,R,C,90,2,1,200,4.0,15.0,6,125,25,1,1.00,0.67,49.120253
9,Bran Flakes,P,C,90,3,0,210,5.0,13.0,5,190,25,3,1.00,0.67,53.313813


In [103]:
df.tail(7)

,name,mfr,type,calories,protein,fat,sodium,fiber,carbo,sugars,potass,vitamins,shelf,weight,cups,rating
70,Total Raisin Bran,G,C,140,3,1,190,4.0,15.0,14,230,100,3,1.5,1.00,28.592785
71,Total Whole Grain,G,C,100,3,1,200,3.0,16.0,3,110,100,3,1.0,1.00,46.658844
72,Triples,G,C,110,2,1,250,0.0,21.0,3,60,25,3,1.0,0.75,39.106174
73,Trix,G,C,110,1,1,140,0.0,13.0,12,25,25,2,1.0,1.00,27.753301
74,Wheat Chex,R,C,100,3,1,230,3.0,17.0,3,115,25,1,1.0,0.67,49.787445
75,Wheaties,G,C,100,3,1,200,3.0,17.0,3,110,25,1,1.0,1.00,51.592193
76,Wheaties Honey Gold,G,C,110,2,1,200,1.0,16.0,8,60,25,1,1.0,0.75,36.187559


In [110]:
df[-1:]

,name,mfr,type,calories,protein,fat,sodium,fiber,carbo,sugars,potass,vitamins,shelf,weight,cups,rating
76,Wheaties Honey Gold,G,C,110,2,1,200,1.0,16.0,8,60,25,1,1.0,0.75,36.187559


In [114]:
df[:1]

,name,mfr,type,calories,protein,fat,sodium,fiber,carbo,sugars,potass,vitamins,shelf,weight,cups,rating
0,100% Bran,N,C,70,4,1,130,10.0,5.0,6,280,25,3,1.0,0.33,68.402973


Exercício 10 — Desafio 🎯

Use todos os métodos aprendidos para fazer uma análise exploratória completa do dataset e escreva um comentário no código resumindo o que você descobriu sobre os cereais. Responda:

Quantos cereais tem no dataset?
        - 2 Cereais. 

Quantos fabricantes diferentes existem?
        - 7

Qual é a média de calorias?
        - 106.88

Tem dados faltando?
        - Não.

Qual cereal tem mais açúcar?
        - Golden Crisp

Qual cereal tem menos calorias?
        - Quaker Oatmeal

In [120]:
df["type"].nunique()

2

In [122]:
df["mfr"].unique()

array(['N', 'Q', 'K', 'R', 'G', 'P', 'A'], dtype=object)

In [126]:
with pd.option_context("float_format", "{:.2f}".format):
    display(df.describe())

,calories,protein,fat,sodium,fiber,carbo,sugars,potass,vitamins,shelf,weight,cups,rating
count,77.00,77.00,77.00,77.00,77.00,77.00,77.00,77.00,77.00,77.00,77.00,77.00,77.00
mean,106.88,2.55,1.01,159.68,2.15,14.60,6.92,96.08,28.25,2.21,1.03,0.82,42.67
std,19.48,1.09,1.01,83.83,2.38,4.28,4.44,71.29,22.34,0.83,0.15,0.23,14.05
min,50.00,1.00,0.00,0.00,0.00,-1.00,-1.00,-1.00,0.00,1.00,0.50,0.25,18.04
25%,100.00,2.00,0.00,130.00,1.00,12.00,3.00,40.00,25.00,1.00,1.00,0.67,33.17
50%,110.00,3.00,1.00,180.00,2.00,14.00,7.00,90.00,25.00,2.00,1.00,0.75,40.40
75%,110.00,3.00,2.00,210.00,3.00,17.00,11.00,120.00,25.00,3.00,1.00,1.00,50.83
max,160.00,6.00,5.00,320.00,14.00,23.00,15.00,330.00,100.00,3.00,1.50,1.50,93.70


In [127]:
df.isnull().sum()

name        0
mfr         0
type        0
calories    0
protein     0
fat         0
sodium      0
fiber       0
carbo       0
sugars      0
potass      0
vitamins    0
shelf       0
weight      0
cups        0
rating      0
dtype: int64

In [136]:
indice = df["sugars"].idxmax()

In [ ]:
df["name"][indice]

'Golden Crisp'

In [138]:
indice_min = df["sugars"].idxmin()

In [139]:
df["name"][indice_min]

'Quaker Oatmeal'